### PR TITLE
Fill in directories for files added by filter.

### LINKED ISSUE #2843: [jib-cli] Correctly create parent directories on filter files
https://github.com/GoogleContainerTools/jib/pull/2700#discussion_r471560251

currently for filter files, the parent directory is ignored by nature of the filter. Because of this, the filtered files are not added with their parents populated. This could lead to an issue where the directory permissions are not as expected.

for a directory structure like:
```
dir1/dir2/
  - file1
  - file2
  - ignore1
```
and a file filter like `**/file*`, the layer parser only adds
```
dir1/dir2/file1
dir1/dir2/file2
```

What it should add is:
```
dir1
dir1/dir2
dir1/dir2/file1
dir1/dir2/file2
```
so that it correctly adds the permissions of `dir1` and `dir2` based on the spec from the build file for directories

### PR DESCRIPTION
**Pull request by loosebazooka**
This happens where parent directories are filtered
but need to be filled in for files in those
directories

fixes #2843

### CODE REVIEW

File: jib-cli/src/main/java/com/google/cloud/tools/jib/cli/buildfile/Layers.java [hunk 1]
```diff
@@ -96,49 +104,72 @@
...
+                } else if (Files.isRegularFile(path)) {
+                  if (!path.startsWith(src)) {
+                    // if we end up in a situation where the file added is somehow outside of the
+                    // tree then we do not know how to properly handle it at the moment. It could
+                    // be from a link scenario that we do not understand.
+                    throw new IllegalStateException(
+                        src.toString() + " is not a parent of " + path.toString());
+                  }
+                  Path parent = path.getParent();
+                  while (true) {
```
Comment (loosebazooka) [inline jib-cli/src/main/java/com/google/cloud/tools/jib/cli/buildfile/Layers.java:156]: Actually these need to be added in reverse order

File: jib-cli/src/main/java/com/google/cloud/tools/jib/cli/buildfile/Layers.java [hunk 2]
```diff
@@ -71,6 +75,10 @@
           Path src = rawSrc.isAbsolute() ? rawSrc : buildRoot.resolve(rawSrc);
           AbsoluteUnixPath dest = copySpec.getDest();
 
+          if (!Files.isDirectory(src) && !Files.isRegularFile(src)) {
```
Comment (chanseokoh) [inline jib-cli/src/main/java/com/google/cloud/tools/jib/cli/buildfile/Layers.java:78]: Is this primarily to guard against symlinks?

Comment (loosebazooka) [reply jib-cli/src/main/java/com/google/cloud/tools/jib/cli/buildfile/Layers.java:78]: Yeah there's two checks. One for non existence and one for unknown types. Perhaps the messaging should be different

File: jib-cli/src/test/resources/buildfiles/layers/includesExcludesTest/layers.yaml [hunk 3]
```diff
@@ -33,11 +33,20 @@ entries:
         excludes:
           # equivalent to "**/excludedDir/**"
           - "**/excludedDir/"
+  - name: "exclude dir and contents"
+    files:
+      - src: "project"
+        dest: "/target/edac"
+        excludes:
+          # equivalent to "**/excludedDir/**"
```
Comment (chanseokoh) [inline jib-cli/src/test/resources/buildfiles/layers/includesExcludesTest/layers.yaml:41]:
```suggestion
```

**Review by chanseokoh — APPROVED** [review]

**Pull request merged**